devices/src/virtio/block.rs

```diff
@@ -325,59 +326,75 @@ impl BlockEpollHandler {
             METRICS.block.event_fails.inc();
+            Err(DeviceErr::FailedSignalingUsedQueue(e))
+        } else {
+            Ok(())
         }
     }
 
-    fn update_disk_image(&mut self, disk_image: File) {
+    fn update_disk_image(&mut self, disk_image: File) -> std::result::Result<(), DeviceErr> {
         self.disk_image = disk_image;
         self.disk_image_id = build_disk_image_id(&self.disk_image);
         METRICS.block.update_count.inc();
+        Ok(())
     }
 }
 
 impl EpollHandler for BlockEpollHandler {
-    fn handle_event(&mut self, device_event: DeviceEventT, _: u32, payload: EpollHandlerPayload) {
+    fn handle_event(
+        &mut self,
+        device_event: DeviceEventT,
+        _: u32,
+        payload: EpollHandlerPayload,
+    ) -> std::result::Result<(), DeviceErr> {
         match device_event {
             QUEUE_AVAIL_EVENT => {
                 METRICS.block.queue_event_count.inc();
                 if let Err(e) = self.queue_evt.read() {
                     error!("Failed to get queue event: {:?}", e);
                     METRICS.block.event_fails.inc();
-                    return;
+                    Err(DeviceErr::FailedReadingQueue {
+                        event_type: "queue event",
+                        underlying: e,
+                    })
                 }
-
                 // While limiter is blocked, don't process any more requests.
-                if self.rate_limiter.is_blocked() {
-                    return;
-                }
-
-                if self.process_queue(0) {
-                    self.signal_used_queue();
+                else if self.rate_limiter.is_blocked() {
+                    Ok(())
+                } else if self.process_queue(0) {
+                    self.signal_used_queue()
+                } else {
+                    Ok(())
                 }
             }
             RATE_LIMITER_EVENT => {
                 METRICS.block.rate_limiter_event_count.inc();
                 // Upon rate limiter event, call the rate limiter handler
                 // and restart processing the queue.
                 if self.rate_limiter.event_handler().is_ok() && self.process_queue(0) {
-                    self.signal_used_queue();
+                    self.signal_used_queue()
+                } else {
+                    Ok(())
                 }
             }
             FS_UPDATE_EVENT => {
                 if let EpollHandlerPayload::DrivePayload(file) = payload {
-                    self.update_disk_image(file);
+                    self.update_disk_image(file)
                 } else {
                     // This path can only be reached if we have a logical problem in our code.
                     panic!("Received update disk image event with empty payload.")
                 }
             }
-            _ => panic!("Unknown event type was received."),
+            unknown => Err(DeviceErr::UnknownEvent {
+                device: "BlockEpollHandler",
```

Comment:
`BlockEpollHandler` isn't an actual device, the device in this case would have to be a `block`.